### PR TITLE
Refactoring IDE models.

### PR DESCRIPTION
**Pull request by farmaazon**
### Pull Request Description
The goal is to mark clear responsibility of models and controllers in ide crate. Also, to make tests simpler, for each model the API trait was added.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md).
- [x] All code has been tested where possible.
- [x] All code has been profiled where possible.

### CODE REVIEW

File: src/rust/ide/tests/language_server.rs [hunk 1]
```diff
@@ -304,27 +303,27 @@ async fn binary_protocol_test() {
     let _guard   = ide::ide::setup_global_executor();
     let project  = setup_project().await;
     println!("Got project: {:?}",project);
-    let path     = Path::new(project.language_server_rpc.content_root(), &["test_file.txt"]);
-    let contents = "Hello!".as_bytes();
-    let written  = project.language_server_bin.write_file(&path,contents).await.unwrap();
+    let path     = Path::new(project.json_rpc().content_root(), &["test_file.txt"]);
+    let contents = "Hello!".to_string();
+    let written  = project.json_rpc().write_file(&path,&contents).await.unwrap();
```
Comment (mwu-tow) [inline src/rust/ide/tests/language_server.rs:308]: not binary anymore

File: src/rust/ide/lib/parser/src/wsclient.rs [hunk 2]
```diff
@@ -197,6 +197,7 @@ impl Client {
     }
 
     pub fn parse(&mut self, program:String, ids:IdMap) -> api::Result<Ast> {
+        println!("Without metadata:\n{}",program);
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/src/wsclient.rs:200]: to remove

File: src/rust/ide/lib/parser/src/wsclient.rs [hunk 3]
```diff
@@ -207,6 +208,7 @@ impl Client {
 
     pub fn parse_with_metadata<M:Metadata>
     (&mut self, program:String) -> api::Result<ParsedSourceFile<M>> {
+        println!("With metadata:\n{}",program);
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/src/wsclient.rs:211]: to remove

File: src/rust/ide/src/controller/graph.rs [hunk 4]
```diff
@@ -783,127 +778,102 @@ mod tests {
...
-            let id     = Id {crumbs};
-            let graph  = Handle::new(Logger::new("Test"),module,parser,id).unwrap();
-            let uid    = graph.all_node_infos().unwrap()[0].id();
-
-            graph.module.with_node_metadata(uid, |data| data.position = Some(pos));
-
+        Fixture::set_up().run(|graph| async move {
+            let uid = graph.all_node_infos().unwrap()[0].id();
+            let pos = model::module::Position {vector:Vector2::new(0.0,0.0)};
+            graph.module.with_node_metadata(uid, Box::new(|data| data.position = Some(pos))); // TODO
```
Comment (mwu-tow) [inline src/rust/ide/src/controller/graph.rs:856]: remove TODO, as apparently there is no neat solution / or think once again

File: src/rust/ide/src/controller/module.rs [hunk 5]
```diff
@@ -178,8 +178,8 @@ impl Handle {
     ) -> FallibleResult<Self> {
         let logger = Logger::new("Mocked Module Controller");
         let ast    = parser.parse(code.to_string(),id_map)?.try_into()?;
-        let model  = model::Module::new(ast, default());
-        let model  = model::synchronized::Module::mock(path,model);
+        let metadata = default();
+        let model  = Rc::new(model::module::Plain::new(path,ast,metadata));
```
Comment (mwu-tow) [inline src/rust/ide/src/controller/module.rs:182]: align

File: src/rust/ide/lib/enso-protocol/src/lib.rs [hunk 6]
```diff
@@ -33,14 +33,16 @@ pub mod prelude {
     pub use logger::enabled::Logger;
 
     pub use std::future::Future;
+    /// We always use local futures in our single-threaded environment
+    pub use futures::future::LocalBoxFuture as BoxFuture;
```
Comment (mwu-tow) [inline src/rust/ide/lib/enso-protocol/src/lib.rs:37]: to reconsider, not consistent

File: src/rust/ide/lib/parser/src/wsclient.rs (hunk 2 above)
Comment (farmaazon) [inline src/rust/ide/lib/parser/src/wsclient.rs:200]: to remove

File: src/rust/ide/lib/parser/src/wsclient.rs (hunk 3 above)
Comment (farmaazon) [inline src/rust/ide/lib/parser/src/wsclient.rs:211]: to remove

File: src/rust/ide/src/model.rs [hunk 7]
```diff
@@ -1,24 +1,39 @@
...
 pub use project::Project;
 pub use suggestion_database::SuggestionDatabase;
+
+/// A module with commonly used traits to mass import.
+pub mod traits {
+    use super::*;
+
+    pub use execution_context::API as TRAIT_ExecutionContextApi;
+    pub use module::API            as TRAIT_ModuleApi;
+    pub use project::API           as TRAIT_ProjectApi;
```
Comment (mwu-tow) [inline src/rust/ide/src/model.rs:38]: Do we have any convention justifying this?

File: src/rust/ide/src/model/execution_context.rs [hunk 8]
```diff
@@ -249,129 +236,65 @@ pub struct AttachedVisualization {
...
-        }
-    }
+    /// Get the registry of computed values.
+    fn computed_value_info_registry(&self) -> &ComputedValueInfoRegistry;
+
+    /// Get all items on stack.
+    fn stack_items<'a>(&'a self) -> Box<dyn Iterator<Item=LocalCall> + 'a>;
+
+    /// Push a new stack item to execution context.
+    fn push(&self, stack_item:LocalCall) -> BoxFuture<FallibleResult<()>>;
```
Comment (mwu-tow) [inline src/rust/ide/src/model/execution_context.rs:261]: recheck if any of methods here could sensibly use static boxed future

File: src/rust/ide/src/controller/graph.rs [hunk 9]
```diff
@@ -783,127 +778,102 @@ mod tests {
...
-        /// Creates module and graph controllers.
-        pub fn create_controllers(&self) -> (controller::Module,Handle) {
-            let ls = language_server::Connection::new_mock_rc(default());
-            self.create_controllers_with_ls(ls)
+        pub fn module_data(&self) -> model::module::test::MockData {
+            model::module::test::MockData {
+                code : self.code.clone(),
+                path : self.module_path.clone(),
+                ..default()
+            }
```
Comment (farmaazon) [inline src/rust/ide/src/controller/graph.rs:807]: I wonder... perhaps there should be rather someting like "MockData" inheriting? Or just one MockData for all structures?

File: src/rust/ide/src/controller/graph.rs [hunk 10]
```diff
@@ -783,127 +778,102 @@ mod tests {
...
 
-        pub fn run<Fut>
-        ( &mut self
-        , data : MockData
-        , test : impl FnOnce(controller::Module,Handle) -> Fut + 'static
-        ) where Fut : Future<Output=()> {
-            let (module,graph) = data.create_controllers();
+        pub fn run<Test,Fut>(&mut self, test:Test)
+            where Test : FnOnce(Handle) -> Fut + 'static,
+                  Fut  : Future<Output=()> {
```
Comment (farmaazon) [inline src/rust/ide/src/controller/graph.rs:843]:
```suggestion
        where Test : FnOnce(Handle) -> Fut + 'static,
              Fut  : Future<Output=()> {
```

File: src/rust/ide/src/controller/graph.rs (hunk 4 above)
Comment (farmaazon) [inline src/rust/ide/src/controller/graph.rs:856]: TODO left.

File: src/rust/ide/src/controller/graph/executed.rs [hunk 11]
```diff
@@ -217,18 +215,15 @@ mod tests {
...
+        let graph_data     = controller::graph::tests::MockData::new();
+        let graph          = graph_data.graph();
+        let execution_data = model::execution_context::plain::test::MockData::new();
+        let execution      = Rc::new(execution_data.create());
+        let project        = Rc::new(model::project::MockAPI::new());
+        let executed_graph = Handle::new_internal(graph,project,execution.clone_ref());
 
         // Generate notification.
-        let notification = execution_data.mock_values_computed_update();
+        let notification = model::execution_context::synchronized::test::Fixture::mock_values_computed_update(&execution_data);
```
Comment (farmaazon) [inline src/rust/ide/src/controller/graph/executed.rs:226]: Too long, let's make an alias import.

File: src/rust/ide/src/controller/searcher.rs [hunk 12]
```diff
@@ -246,9 +246,9 @@ impl Searcher {
             data            : default(),
             notifier        : default(),
             module          : Rc::new(project.qualified_module_name(&module)),
-            database        : project.suggestion_db.clone_ref(),
-            language_server : project.language_server_rpc.clone_ref(),
-            parser          : project.parser.clone_ref(),
+            database        : project.suggestion_db(),
+            language_server : project.json_rpc(),
+            parser          : project.parser().clone_ref(),
```
Comment (farmaazon) [inline src/rust/ide/src/controller/searcher.rs:251]: Now I think, that parser should also return owned value (as suggestion_db and json_rpc do).

File: src/rust/ide/src/controller/text.rs [hunk 13]
```diff
@@ -65,7 +65,7 @@ impl Handle {
         } else {
             FileHandle::PlainText {
                 path            : Rc::new(path),
-                language_server : project.language_server_rpc.clone_ref()
+                language_server : project.json_rpc().clone_ref()
```
Comment (farmaazon) [inline src/rust/ide/src/controller/text.rs:68]:
```suggestion
                language_server : project.json_rpc()
```

File: src/rust/ide/src/model.rs [hunk 14]
```diff
@@ -1,24 +1,39 @@
...
+//! * Sometimes a Mock generated with `mock-all` crate.
+//! * A Shared Handle, which hides some of above API implementors under `Rc<dyn API>`.
 //!
-//! The models should implement internal mutability pattern, because they should be easily shared
-//! between controllers.
+//! Plain Model and Mock should be used in tests only, which simplifies mocking dependencies. In
+//! the actual code the Synchronized Model should be used. Each struct using model as a dependency
+//! should use the Shared Handle.
+//!
+//! Due to Shared Handle object, all models should implement "internal mutability" pattern.
```
Comment (farmaazon) [inline src/rust/ide/src/model.rs:19]: Let's rephrase that, like "As the models are usually shared between many controllers, they should implement "internal mutability" pattern."

File: src/rust/ide/src/model/module/synchronized.rs [hunk 15]
```diff
@@ -169,7 +226,7 @@ impl Module {
     /// so any new update attempt should perform full invalidation.
     fn new_ls_content_info
     (&self, old_content:ContentSummary, new_content:FallibleResult<ParsedContentSummary>)
-    -> LanguageServerContent {
+     -> LanguageServerContent {
```
Comment (farmaazon) [inline src/rust/ide/src/model/module/synchronized.rs:229]:
```suggestion
    -> LanguageServerContent {
```

File: src/rust/ide/src/model/module/synchronized.rs [hunk 16]
```diff
@@ -183,7 +240,7 @@ impl Module {
     /// content summery of Language Server state.
     async fn handle_notification
     (&self, content:&LanguageServerContent, notification:Notification)
-    -> FallibleResult<ParsedContentSummary> {
+     -> FallibleResult<ParsedContentSummary> {
```
Comment (farmaazon) [inline src/rust/ide/src/model/module/synchronized.rs:243]:
```suggestion
    -> FallibleResult<ParsedContentSummary> {
```

File: src/rust/ide/src/model/module/synchronized.rs [hunk 17]
```diff
@@ -227,13 +284,13 @@ impl Module {
     /// Language Server. Returns the new summary of Language Server state.
     async fn notify_language_server
     ( &self
-    , ls_content        : &ContentSummary
-    , edits_constructor : impl FnOnce(SourceFile) -> Vec<TextEdit>
+      , ls_content        : &ContentSummary
+      , edits_constructor : impl FnOnce(SourceFile) -> Vec<TextEdit>
```
Comment (farmaazon) [inline src/rust/ide/src/model/module/synchronized.rs:288]:
```suggestion
    , ls_content        : &ContentSummary
    , edits_constructor : impl FnOnce(SourceFile) -> Vec<TextEdit>
```

File: src/rust/ide/src/model/module/synchronized.rs [hunk 18]
```diff
@@ -320,7 +377,7 @@ pub mod test {
         }
 
         fn expect_edit<EditApplier>(&self, result:json_rpc::Result<()>, edit_applier:EditApplier)
-        where EditApplier : FnOnce(&[TextEdit]) -> String + 'static {
+            where EditApplier : FnOnce(&[TextEdit]) -> String + 'static {
```
Comment (farmaazon) [inline src/rust/ide/src/model/module/synchronized.rs:380]:
```suggestion
        where EditApplier : FnOnce(&[TextEdit]) -> String + 'static {
```

File: src/rust/ide/src/view/layout.rs [hunk 19]
```diff
@@ -93,9 +93,10 @@ impl ViewLayout {
     , text_controller          : controller::Text
     , graph_controller         : controller::ExecutedGraph
     , visualization_controller : controller::Visualization
-    , project                  : Rc<model::Project>
+    , project                  : impl Into<Rc<dyn model::project::API>> // TODO[ao]!!!!!!!!!!!!
```
Comment (farmaazon) [inline src/rust/ide/src/view/layout.rs:96]: Uhm, well, left TODO

File: src/rust/ide/src/view/node_searcher.rs [hunk 20]
```diff
@@ -37,7 +37,7 @@ impl NodeSearcher {
     , node_editor   : NodeEditor
     , fonts         : &mut font::Registry
     , focus_manager : &FocusManager
-    , project       : Rc<model::Project>)
+    , project       : model::Project)
```
Comment (farmaazon) [inline src/rust/ide/src/view/node_searcher.rs:40]: ) should be in a new line.

File: src/rust/ide/src/model/execution_context/synchronized.rs [hunk 21]
```diff
@@ -0,0 +1,420 @@
...
+    // /// Create a mock which does no call on `language_server` during construction.
+    // #[cfg(test)]
+    // fn new_mock
+    // ( id              : model::execution_context::Id
+    // , model           : model::ExecutionContext
+    // , language_server : Rc<language_server::Connection>
+    // ) -> Self {
+    //     let logger          = Logger::new("ExecuctionContext mock");
+    //     ExecutionContext {id,model,language_server,logger}
+    // }
```
Comment (mwu-tow) [inline src/rust/ide/src/model/execution_context/synchronized.rs:106]: to be removed?

File: src/rust/ide/src/model/project/synchronized.rs [hunk 22]
```diff
@@ -0,0 +1,486 @@
...
+        let event = language_server::Event::Notification(language_server::Notification::ExpressionValuesComputed(notification.clone()));
+        json_events_sender.unbounded_send(event).unwrap();
+        test.run_until_stalled();
+
+        // Context now has the information about type.
+        let value_info = value_registry.get(&expression_id).unwrap();
+        assert_eq!(value_info.typename, value_update.typename.clone().map(ImString::new));
+        assert_eq!(value_info.method_pointer, value_update.method_call.clone().map(Rc::new));
+    }
+}
```
Comment (mwu-tow) [inline src/rust/ide/src/model/project/synchronized.rs:486]: nl

File: src/rust/ide/src/model.rs (hunk 7 above)
Comment (farmaazon) [reply src/rust/ide/src/model.rs:38]: https://github.com/enso-org/ide/blob/main/docs/contributing/style-guide.md#trait-exporting

File: src/rust/ide/src/controller/graph.rs (hunk 9 above)
Comment (mwu-tow) [reply src/rust/ide/src/controller/graph.rs:807]: Agreed to leave this for future improvement.

File: src/rust/ide/src/controller/graph.rs (hunk 4 above)
Comment (mwu-tow) [reply src/rust/ide/src/controller/graph.rs:856]: Agreed to not take action here at this time. TODO will be removed.

File: src/rust/ide/src/model/execution_context.rs (hunk 8 above)
Comment (mwu-tow) [reply src/rust/ide/src/model/execution_context.rs:261]: To reduce possible future breakages we decided to leave the API as-is.
Could be revisited if use case appears.

File: src/rust/ide/src/view/layout.rs (hunk 19 above)
Comment (mwu-tow) [reply src/rust/ide/src/view/layout.rs:96]: use `model::Project`
and link to issue https://github.com/rust-lang/rust/issues/63033

**Pull request merged**